libs/wingc/src/lifting.rs

```diff
@@ -0,0 +1,360 @@
+			}
+			_ => false,
+		}
+	}
+
+	fn jsify_expr(&self, node: &Expr, phase: Phase) -> String {
+		self.jsify.jsify_expression(
+			&node,
+			&mut JSifyContext {
+				in_json: false,
+				phase: phase,
+				files: &mut Files::default(),
+				lifts: &mut Lifts::disabled(),
+			},
+		)
+	}
+}
+
+impl<'a> Fold for LiftTransform<'a> {
+	fn fold_reference(&mut self, node: Reference) -> Reference {
+		match node {
+			Reference::InstanceMember {
+				object,
+				property,
+				optional_accessor,
+			} => {
+				self.ctx.push_property(property.name.clone());
+				let result = Reference::InstanceMember {
+					object: Box::new(self.fold_expr(*object)),
+					property: self.fold_symbol(property),
+					optional_accessor,
+				};
+				self.ctx.pop_property();
+				return result;
+			}
+			Reference::TypeMember { typeobject, property } => {
+				self.ctx.push_property(property.name.clone());
+				let result = Reference::TypeMember {
+					typeobject: Box::new(self.fold_expr(*typeobject)),
+					property: self.fold_symbol(property),
+				};
+				self.ctx.pop_property();
+				return result;
+			}
+			_ => {}
+		}
+
+		fold::fold_reference(self, node)
+	}
+
+	fn fold_expr(&mut self, node: Expr) -> Expr {
+		let expr_phase = self.jsify.types.get_expr_phase(&node).unwrap();
+		let expr_type = self.jsify.types.get_expr_type(&node).unwrap();
+		let is_field = includes_this(&node);
+
+		// this whole thing only applies to inflight expressions
+		if self.ctx.current_phase() == Phase::Preflight {
+			return fold::fold_expr(self, node);
+		}
+
+		// if this expression represents the current class, no need to capture it (it is by definition
+		// available in the current scope)
+		if self.is_self_type_reference(&node) {
+			return fold::fold_expr(self, node);
+		}
+
+		//---------------
+		// LIFT
+
+		if expr_phase == Phase::Preflight {
+			// jsify the expression so we can get the preflight code
+			let code = self.jsify_expr(&node, Phase::Preflight);
+
+			// check that we can qualify the lift (e.g. determine which property is being accessed)
+			if self.ctx.current_property().is_none()
+				&& expr_type.as_preflight_class().is_some()
+				&& !expr_type.is_handler_preflight_class()
+			{
+				report_diagnostic(Diagnostic {
+					message: format!(
+						"Cannot qualify access to a lifted object of type \"{}\"",
+						expr_type.to_string()
+					),
+					span: Some(node.span.clone()),
+				});
+
+				return node;
+			}
+
+			// println!("Lifting: {}", preflight);
+			let mut lifts = self.lifts_stack.pop().unwrap();
+			lifts.lift(
+				node.id,
+				self.ctx.current_method(),
+				self.ctx.current_property(),
+				is_field,
+				&code,
+			);
+			self.lifts_stack.push(lifts);
+
+			return node;
+		}
+
+		//---------------
+		// CAPTURE
+
+		if !is_field && self.should_capture_expr(&node) {
+			// jsify the expression so we can get the preflight code
+			let preflight = self.jsify_expr(&node, Phase::Inflight);
+
+			// println!("Capturing: {}", preflight);
```

Comment:
```suggestion
```
note the use of `debug!` macros in the jsii_importer (found it useful in the past)